### PR TITLE
Delegate multi-dataset tuning to MultiTrainer

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

- replace duplicated Tuner dataset orchestration with the existing MultiTrainer owner
- keep every dataset training run in an isolated YOLO CLI subprocess
- preserve YOLOWorld and YOLOE filename dispatch through the temporary CLI checkpoint
- atomically assign exact defaults to the first MongoDB worker and mutate defaults on concurrent workers
- store per-dataset metrics, macro-mean metrics, and cleanup paths in each MongoDB result document
- identify distributed winners by their result paths instead of a worker-local loop index
- bump the package version to 8.4.129

The final diff changes only three existing files and is net-negative: 36 additions, 63 deletions.

## Validation

- final-head single-dataset Mongo Tuner run: 1 epoch, 1 generation, exact defaults, CLI child, mean/individual metrics, checkpoints, plots, YAML, NDJSON, and clean process exit
- final-head five-dataset Mongo canary: 5 epochs, 3 generations, 15/15 dataset runs completed in 509.39 seconds
    - generation 1 exactly matched every default hyperparameter; generations 2-3 mutated
    - all three documents contain five non-empty per-dataset metric dictionaries and verified macro means
    - all three documents retain five `save_dirs`; cleanup kept only the best generation's five dataset directories
    - five best/last checkpoint pairs, both Tuner plots, best YAML, three-row NDJSON, and zero temporary base models verified
- full UL33 Mongo canary on the CLI/MultiTrainer path: 5 epochs, 3 generations, 99/99 dataset runs completed with 33 metric dictionaries per document
- real two-worker empty-collection check: exactly one worker claimed defaults, the other mutated, and the fitness index was created
- real two-dataset Tuner smoke including a duplicate dataset name
- verified a `yolov8s-world.yaml` source produces a `yolov8s-world.pt` CLI checkpoint path
- `uvx ruff format --check ultralytics/engine/tuner.py ultralytics/engine/trainer.py ultralytics/__init__.py`
- `uvx ruff check --ignore BLE001 ultralytics/engine/tuner.py ultralytics/engine/trainer.py ultralytics/__init__.py`
- `python3 -m compileall -q ultralytics/engine/tuner.py ultralytics/engine/trainer.py ultralytics/__init__.py`
- `git diff --check origin/main...HEAD`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Multi-dataset tuning now delegates training and metric collection to `MultiTrainer`, reducing duplicated orchestration while preserving per-dataset results and cleanup tracking.

### 📊 Key Changes

- `Tuner.__call__` creates a `YOLO` model and uses `MultiTrainer` to run all datasets, collect individual metrics, compute mean metrics, and locate dataset output directories.
- `MultiTrainer` now exposes `mean_metrics` and derives temporary base checkpoint names from the configured model name.
- MongoDB result documents now include per-dataset `metrics` and `save_dirs`; collection creation and fitness-index setup coordinate concurrent workers so only one claims the initial defaults.
- Best-result detection compares saved dataset paths instead of relying on the current worker's iteration index.
- The package version was bumped from `8.4.128` to `8.4.129`.

### 🎯 Purpose & Impact

- Multi-dataset tuning uses the existing `MultiTrainer` workflow, while MongoDB records retain the metrics and paths needed to identify winners and remove non-winning dataset outputs.
- Tuning results continue to include aggregate fitness alongside individual dataset metrics, with winner selection no longer dependent on worker-local iteration numbering.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/tuner.py [hunk 1]
```diff
@@ -488,46 +474,19 @@ def __call__(self, iterations: int = 10, cleanup: bool = True):
...
             if not isinstance(data, (list, tuple)):
                 data = [data]
-            dataset_names = self._dataset_names(data)
-            save_dir = (
-                [get_save_dir(get_cfg(train_args))]
-                if len(data) == 1
-                else [get_save_dir(get_cfg(train_args), name=name) for name in dataset_names]
-            )
+            model = YOLO(train_args.pop("model"))
+            trainer = MultiTrainer(None, {**train_args, "data": data}, model.model)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/tuner.py:488]: ⚠️ **HIGH**: This loses specialized model dispatch for YOLOWorld and YOLOE. The parent `YOLO(...)` resolves those wrappers from the original filename, but `MultiTrainer` serializes the model as `multitrain_base.pt`; each child CLI then no longer sees `world`/`yoloe` in the model name and constructs the generic `YOLO` wrapper, selecting standard trainers instead of `WorldTrainer`/`YOLOETrainer`. Tuning these models can therefore fail or run with incorrect training behavior. Preserve the specialized wrapper/type when creating the subprocess checkpoint or make the child dispatch inspect checkpoint model type.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The delegation is directionally correct, but the current change regresses specialized YOLOWorld/YOLOE dispatch and crashes on mixed dataset success/failure results. These issues should be fixed before merging.

<details><summary>💬 Posted 2 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/engine/tuner.py:478` This loses specialized model dispatch for YOLOWorld and YOLOE. The parent `YOLO(...)` resolves those wrappers from the original filename, but `MultiTrainer` serializes the model as `multitrain_base.pt`; each child CLI then no longer sees `world`/`yoloe` in the model name and constructs the generic `YOLO` wrapper, selecting standard trainers instead of `WorldTrainer`/`YOLOETrainer`. Tuning these models can therefore fail or run with incorrect training behavior. Preserve the specialized wrapper/t…
- 💡 **MEDIUM** `ultralytics/engine/tuner.py:479` `MultiTrainer.train()` records failed datasets as `None`, whereas the removed loop stored `{}`. For a mixed success/failure iteration, `_best_result_index()` can select this result, and `_best_metrics()` then calls `.get()` on every multi-dataset value, raising `AttributeError` on the `None` entry and aborting tuning during result logging. Normalize failed metrics to empty dictionaries before passing them into the tuner result structures.
</details>

File: ultralytics/engine/tuner.py [hunk 2]
```diff
@@ -488,46 +474,19 @@ def __call__(self, iterations: int = 10, cleanup: bool = True):
...
+            dataset_metrics = trainer.train()
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/tuner.py]: 💡 **MEDIUM**: `MultiTrainer.train()` records failed datasets as `None`, whereas the removed loop stored `{}`. For a mixed success/failure iteration, `_best_result_index()` can select this result, and `_best_metrics()` then calls `.get()` on every multi-dataset value, raising `AttributeError` on the `None` entry and aborting tuning during result logging. Normalize failed metrics to empty dictionaries before passing them into the tuner result structures.

**Suggested change:**
```suggestion
            dataset_metrics = {dataset: metrics or {} for dataset, metrics in trainer.train().items()}
```

File: ultralytics/engine/tuner.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/engine/tuner.py:488]: Fixed in 50740d0e1. MultiTrainer now preserves the original model stem in its temporary .pt checkpoint filename, so YOLOWorld and YOLOE CLI dispatch survives serialization. Focused validation with model=yolov8s-world.yaml captured a child CLI model path ending in yolov8s-world.pt and verified temporary checkpoint cleanup.

File: ultralytics/engine/tuner.py (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics/engine/tuner.py]: Fixed in 50740d0e1 using the suggested normalization. Failed MultiTrainer metrics are converted from None to the existing Tuner empty-dict contract before result selection, logging, Mongo persistence, and plotting.

**Comment by glenn-jocher** [comment]
Cold full-diff review on exact live head `1aca8ef0a399d5dc60da1f733fc3ad48b98a4b6f`: LGTM, no findings. The change is confined to the existing Tuner/MultiTrainer owners plus the patch version, is net-negative, preserves CLI subprocess isolation, and the final-head single- and five-dataset Mongo canaries passed with exact-default claiming, macro/per-dataset metrics, artifacts, and cleanup verified. Unresolved review threads: 0.

**Comment by UltralyticsAssistant** [comment]
🎉 **PR [#25937](https://github.com/ultralytics/ultralytics/pull/25937) has been merged!**

Huge thanks to @glenn-jocher, with contributions from @raimbekovm, @artest08, @Y-T-G, and @JESUSROYETH. This work streamlines multi-dataset tuning through `MultiTrainer` while preserving per-dataset metrics and output paths, improving winner detection and cleanup coordination across workers.

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

A meaningful example of focused improvements coming together to make tuning orchestration more reliable and maintainable.
